cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,576 @@
+    // +optional
+    Resources *ResourceBufferCapacity
+    // +optional
+    Pods *PodBufferCapacity
+}
+
+type ReplicaBufferCapacity struct {
+	// An object implementing scale subresource (for example deployment, job etc.)
+    TargetRef *TypedLocalObjectReference
+    // One of the following must be set:
+    // +optional
+    Exactly *int
+    // +optional
+    Percent *PercentReplicaBufferCapacity
+}
+
+type PercentReplicaBufferCapacity struct {
+    Percent  int
+    MinCount *int
+    MaxCount *int
+}
+
+type ResourceBufferCapacity struct {
+	// Defines where the resource should be created.
+	TargetRef *TypedLocalObjectReference
+    // Upper limit of the total capacity that will be scaled for this buffer.
+    // The total capacity will be split into smaller chunks that each fits onto nodes.
+    // The size of a chunk will be either decided for the user (using some defaults and heuristics) or the
+    // user can specify per pod Requests to define the shape of the buffer space. In the latter case the
+    // CPU and Mem will be the maximum size of the buffer and if the Requests are too big to fit the total
+    // buffer size the admission of the buffer should be denied.
+    TotalCpu resource.Quantity
+    TotalMem resource.Quantity
+
+    // Optional if defining a single chunk size is needed. The controller will translate this into enough
+    // pods to create buffer not exceeding the total size above
+    PerChunkResources *ResourceLimits
+    // Optional if more detailed limitations than target object ref on its own is not enough.
+    NodeSelector NodeSelector
+}
+
+type PodBufferCapacity struct {
+    ReplicaNum int
+    PodTemplate PodTemplateRef
+}
+
+type NodeAutoscalingBufferStatus struct {
+    PodBufferCapacity   PodBufferCapacity
+	PodTemplateGeneration int	
+    Conditions           []BufferCondition
+    LastFullyProvisioned *metav1.Time
+}
+```
+
+### What is the responsibility split between the NodeAutoscalingBuffers controller and autoscaler?
+
+The controller would own the logic of translating the buffer configuration into
+a pod spec that would define what spare space needs to be provisioned in the
+cluster to fulfill the requirements of the buffer. The autoscaler component
+would use the pod spec to provision the capacity.
+
+In case of the cluster autoscaler we would deploy the controller as an optional
+subprocess:
+
+![Figure](./images/buffers-component-structure.jpg)
+
+This design has two main advantages compared to embedding all the logic in the
+library that would be used in autoscaler code:
+
+*   Responsibility split: it splits the responsibility of understanding buffer
+    configuration and exposes to the user what exactly will be provisioned in
+    the cluster based on the buffer spec that was provided. Also if for example
+    the target deployment changes it will be visible whether the change was
+    already picked up.
+*   Extendability: this design will allow for the users to either opt in the
```

Comment:
I'm worried that we are trying to do this but aren't really sure how this would be handled concretely -- there isn't a strong use-case that I can think of for having divergent behavior on the pods that you gen from a given template -- I actually think that this makes things more confusing for a user who has to reason about what kind of output that they will get based on the components that they install